File: native-spring-api/src/test/java/com/xavierbouclet/nativeapi/PostgresSQLExtension.java
package com.xavierbouclet.nativeapi;

import org.junit.jupiter.api.extension.AfterAllCallback;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.springframework.boot.test.util.TestPropertyValues;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ConfigurableApplicationContext;
import org.testcontainers.containers.PostgreSQLContainer;

public class PostgresSQLExtension implements AfterAllCallback {

    private static PostgreSQLContainer<?> postgresSQLServer = new PostgreSQLContainer<>("postgres:14.1")
            .withDatabaseName("integration-tests-db")
            .withUsername("sa")
            .withPassword("sa");


    static {
        postgresSQLServer.start();
    }

    public void afterAll(ExtensionContext p0) {
        postgresSQLServer.stop();
    }



    public static class Initializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {

        public Initializer() {
        }

        @Override
        public void initialize(ConfigurableApplicationContext configurableApplicationContext) {
            TestPropertyValues.of(
                    "spring.r2dbc.url=" + postgresSQLServer.getJdbcUrl().replace("jdbc", "r2dbc"),
                    "spring.r2dbc.username=" + postgresSQLServer.getUsername(),
                    "spring.r2dbc.password=" + postgresSQLServer.getPassword()
            ).applyTo(configurableApplicationContext.getEnvironment());
        }


    }
}

File: native-spring-api/src/test/kotlin/com/lafabrique/digit/owl/PostgresSQLExtension.kt
package com.lafabrique.digit.owl

import org.junit.jupiter.api.extension.AfterAllCallback
import org.junit.jupiter.api.extension.BeforeAllCallback
import org.junit.jupiter.api.extension.ExtensionContext
import org.springframework.boot.test.util.TestPropertyValues
import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.ConfigurableApplicationContext
import org.testcontainers.containers.PostgreSQLContainer


class PostgresSQLExtension : BeforeAllCallback, AfterAllCallback {

    internal class SpecifiedPostgresSQLContainer(imageName: String) : PostgreSQLContainer<SpecifiedPostgresSQLContainer>(imageName)

    companion object {
        private val postgresSQLServer = SpecifiedPostgresSQLContainer("postgres:15.1")
            .withDatabaseName("integration")
            .withUsername("sa")
            .withPassword("sa")
            .waitingFor(
                TestContainerPostgresSQLWaitStrategy()
            )

    }

    override fun beforeAll(p0: ExtensionContext?) {
        postgresSQLServer.start()
        println("Postgres has started")
    }

    override fun afterAll(p0: ExtensionContext?) {
        postgresSQLServer.stop()
    }

    internal class Initializer : ApplicationContextInitializer<ConfigurableApplicationContext> {
        override fun initialize(configurableApplicationContext: ConfigurableApplicationContext) {
            TestPropertyValues.of(
                "spring.datasource.url=${postgresSQLServer.jdbcUrl}",
                "spring.datasource.username=${postgresSQLServer.username}",
                "spring.datasource.password=${postgresSQLServer.password}",
                "spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.PostgreSQLDialect",
                "spring.jpa.hibernate.ddl-auto=update"
            ).applyTo(configurableApplicationContext.environment)
        }
    }
}
